Use contractual limit tracking

TaiWilkin opened the pull request:
## Overview

Previously, a user's number of API calls was tied to the calendar year,
so they could theoretically make 50 calls on Dec 31 and another 50 on
Jan 1. Once we get into paying customers, this becomes more of a
problem. The API limits are now tied to the contract start date, rather
than the calendar year. For free-tier users, we use the date of their
first successful request as the start of their term.

Connects #1410, #1225

## Testing Instructions

* On `develop`, run `./scripts/manage shell_plus`, then enter:
```
from datetime import date
ApiLimit.objects.create(contributor_id=2, yearly_limit=10) #note the id
ApiLimit.objects.create(contributor_id=3, yearly_limit=10)
ApiLimit.objects.create(contributor_id=4, yearly_limit=10)
for x in range(11):
            RequestLog.objects.create(user_id=2, response_code=200)
            r = RequestLog.objects.create(user_id=3, response_code=200)
	    r.created_at = date(2021, 7, 1)
	    r.save()
ApiBlock.objects.create(contributor_id=4, until=date(2021, 12, 31), active=False, limit=10, actual=11) #note the id
ApiBlock.objects.create(contributor_id=5, until=date(2021, 12, 31), active=False, limit=10, actual=11) #note the id
```
* Switch to this branch and run `./scripts/manage migrate`. The migration should succeed.
* Run `./scripts/manage shell_plus`
```
ApiLimit.objects.get(id=1).period_start_date # id matches first limit created. Date should be today
ApiBlock.objects.get(id=1).until # id matches first block created. Date should be today
ApiBlock.objects.get(id=2).until # id matches second block created. The date should still be 12/31/2021 because this block has no associated ApiLimit
```
* Run `./scripts/manage check_api_limits`. An ApiBlock notification should appear.  Run `./scripts/manage shell_plus`
```
ApiBlock.objects.filter(contributor_id=2) # should exist, and "until" is a year from today
ApiBlock.objects.filter(contributor_id=3) # shouldn’t exist since the RequestLogs are all prior to the period_start_date
```
* Run `./scripts/test` and ensure all tests pass

## Checklist

- [x] `fixup!` commits have been squashed
- [x] CI passes after rebase
- [x] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

jwalgran reviewed (approved):
Nice job working implementing this tricky logic. I was able to successfully run the interactive tests in the description and a few of my own. 

I made couple of suggestions inline and would also recommend changing the phrase "for this year" to "per year" in our email notification templates. "For this year" implies more strongly that the limit will reset at the end of the calendar year.

TaiWilkin commented on src/django/api/tests.py:
```diff
@@ -6434,6 +6437,23 @@ def test_under_limit_does_nothing(self):
+            r.created_at = last_month
+            r.save()
+        else:
+            check_api_limits(timezone.now())
```
Updated in 5b3f65a

TaiWilkin commented on src/django/api/limits.py:
```diff
@@ -20,32 +23,45 @@ def get_api_block(contributor):
 @transaction.atomic
 def check_contributor_api_limit(at_datetime, c):
     contributor = Contributor.objects.get(id=c.get('contributor'))
+    logs = c.get('logs')
```
Good call, that's much clearer!

TaiWilkin commented:
> would also recommend changing the phrase "for this year" to "per year" in our email notification templates. "For this year" implies more strongly that the limit will reset at the end of the calendar year.

I went ahead and implemented #1225 which updates the API limit text (in a separate commit) and removed the 'year' language entirely.

jwalgran commented:
Looking at this now.

jwalgran reviewed (approved):
Great idea pulling in #1225. I retested this and it looks good 👍

TaiWilkin commented:
Thanks for reviewing again!
